### PR TITLE
test: exitPool happy path

### PR DESCRIPTION
**Pull request by 0xAustrian**
Closes BAL-31

### CODE REVIEW

File: test/unit/BPool.t.sol [hunk 1]
```diff
@@ -57,6 +57,10 @@ abstract contract BasePoolTest is Test, BConst, Utils {
     bPool.set__records(_token, _record);
   }
 
+  function _setBalance(address _user, uint256 _balance) internal {
+    _writeUintAtAddressMapping(address(bPool), 0, _user, _balance);
+  }
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:62]: please move next to totalSupply, and add the comment on why we're using this method and not smock one

File: test/unit/BPool.t.sol [hunk 2]
```diff
@@ -406,6 +410,73 @@ contract BPool_Unit_JoinPool is BasePoolTest {
...
+          index: 0, // NOTE: irrelevant for this method
+          denorm: 0, // NOTE: irrelevant for this method
+          balance: _fuzz.balance[i]
+        })
+      );
+    }
+
+    // Set LP token balance
+    _setBalance(address(bPool), _fuzz.initPoolSupply); // bPool
+    _setBalance(address(this), _fuzz.initPoolSupply); // fn caller
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:443]: hmm, so the supply is actually bigger? 👀 
i believe the 1st line is irrelevant, bPool balance in bPool isn't used at all

File: test/unit/BPool.t.sol (hunk 1 above)
Comment (wei3erHase) [reply test/unit/BPool.t.sol:62]: i wonder, should we use `vm.deal(address token, address to, uint256 amount)`? i wonder if that increases total supply, could you confirm?

File: test/unit/BPool.t.sol [hunk 3]
```diff
@@ -406,6 +410,73 @@ contract BPool_Unit_JoinPool is BasePoolTest {
...
+  }
+
+  modifier happyPath(ExitPool_FuzzScenario memory _fuzz) {
+    _assumeHappyPath(_fuzz);
+    _setValues(_fuzz);
+    _;
+  }
+
+  function test_HappyPath(ExitPool_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    uint256[] memory minAmountsOut = new uint256[](tokens.length); // Using min possible amounts
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:475]: should we create a `_zeroAmountsArray()` method?

File: test/unit/BPool.t.sol [hunk 4]
```diff
@@ -406,6 +410,73 @@ contract BPool_Unit_JoinPool is BasePoolTest {
...
+    vm.assume(_fuzz.initPoolSupply < type(uint256).max / BONE);
+
+    uint256 _poolAmountInAfterFee = _fuzz.poolAmountIn - (_fuzz.poolAmountIn * EXIT_FEE);
+    vm.assume(_poolAmountInAfterFee <= _fuzz.initPoolSupply);
+    vm.assume(_poolAmountInAfterFee * BONE > _fuzz.initPoolSupply);
+
+    uint256 _ratio = (_poolAmountInAfterFee * BONE) / _fuzz.initPoolSupply; // bdiv uses '* BONE'
+
+    for (uint256 i = 0; i < _fuzz.balance.length; i++) {
+      vm.assume(_fuzz.balance[i] >= BONE); // BONE > MIN_BALANCE so we can do this
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:463]: why is this needed?
wBTC has 8 decimals, we're assuming we need min 1e18?

Comment (0xAustrian) [reply test/unit/BPool.t.sol:466]: I had the 2 assumes `vm.assume(_fuzz.balance[i] >= BONE);` and `vm.assume(_fuzz.balance[i] >= MIN_BALANCE);`. And because BONE is bigger than MIN_BALANCE I deleted the latter, but will double check again.

File: test/unit/BPool.t.sol [hunk 5]
```diff
@@ -69,6 +69,10 @@ abstract contract BasePoolTest is Test, BConst, Utils {
     bPool.set__finalized(_isFinalized);
   }
 
+  function _setBalance(address _user, uint256 _balance, bool _adjustTotalSupply) internal {
+    deal(address(bPool), _user, _balance, _adjustTotalSupply);
+  }
+
   function _setTotalSupply(uint256 _totalSupply) internal {
     // NOTE: not in smock as it uses ERC20.totalSupply()
     _writeUintToStorage(address(bPool), 2, _totalSupply);
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:78]: `deal(to: address(0),...,true)`? 😉

File: test/unit/BPool.t.sol [hunk 6]
```diff
@@ -69,6 +69,10 @@ abstract contract BasePoolTest is Test, BConst, Utils {
     bPool.set__finalized(_isFinalized);
   }
 
+  function _setBalance(address _user, uint256 _balance, bool _adjustTotalSupply) internal {
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:72]: let's not use the boolean and always set it to true, this method should be named so that a reference to being the pool's balance is understood, perhaps `_setPoolBalance`?

in other method we may need `_setPoolTokenBalance` for the erc20 balance of the pool in a token

File: test/unit/BPool.t.sol [hunk 7]
```diff
@@ -406,6 +414,72 @@ contract BPool_Unit_JoinPool is BasePoolTest {
...
+    _assumeHappyPath(_fuzz);
+    _setValues(_fuzz);
+    _;
+  }
+
+  function test_HappyPath(ExitPool_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    uint256[] memory minAmountsOut = _zeroAmountsArray(); // Using min possible amounts
+
+    bPool.exitPool(_fuzz.poolAmountIn, minAmountsOut);
+  }
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:481]: wdyt?
```suggestion
  function test_HappyPath(ExitPool_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
    bPool.exitPool(_fuzz.poolAmountIn, _zeroAmountsArray()); // Using min possible amounts out
  }
```

**Review by wei3erHase — APPROVED** [review]
🫡
